### PR TITLE
[v6r16] RabbitMQ module

### PR DESCRIPTION
**Pull request by Sbalbp**
Created a RabbitMQ module to send and receive messages from queues.

### CODE REVIEW

**Comment by wkrzemien** [comment]
Since Federico asked me for it :), here are some comments to the code:
1.The name of the class is a bit misleading RabbitInterface  suggests that one should inherite from that class and I understood it is not the purpuse.. So I would change the word Interface to something else.
2.i would prefer to define and initialize (by None if needed) all class fields (self.blabla) in the _init_. I don't know what is the common practice in python, by even for readibility, i think its easier to have a look at _init_ and see what are the class variables.
3;It would be very useful to add even a simple use case, like some several-line example to get the idea, how  to use this class.

File: Core/Utilities/RabbitMQ.py [hunk 1]
```diff
@@ -0,0 +1,144 @@
+"""
+Class for management of RabbitMQ connections
+"""
+
+__RCSID__ = "$Id$"
+
+import datetime
```
Comment (wkrzemien) [inline Core/Utilities/RabbitMQ.py]: this is never used

File: Core/Utilities/RabbitMQ.py [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+    if self.receiver:
+      if messageCallback:
+        self.on_message = messageCallback
+      else:
+        self.on_message = self.defaultCallback
+
+    self.queueName = queueName
+    setup = gConfig.getValue( '/DIRAC/Setup', '' )
+
+    result = gConfig.getOption( '/Systems/%s/%s/MessageQueueing/%s/Host' % ( system, setup, queueName ) )
```
Comment (wkrzemien) [inline Core/Utilities/RabbitMQ.py]: All this code up to line 70 seems the same mabye one could refactor by adding some function that would take one string as argument "host", "port" etc

**Comment by Sbalbp** [comment]
In regards to the last comment, I would rather keep that code as it is just for homogeneity with the rest of the code ( there are some other modules in DIRAC with similar cases where nothing was refactored )

**Comment by atsareg** [comment]
What makes this class RabbitMQ specific ? Should you rather make a MQClient that can eventually work with any AMQP service ?

File: Core/Utilities/MessagingQueueConnector.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
+"""
+Abstract class for management of MQ connections
+"""
+
+__RCSID__ = "$Id$"
+
+import time
+from DIRAC import gLogger
+
+class MessagingQueueConnector( object ):
```
Comment (atsareg) [inline Core/Utilities/MessagingQueueConnector.py]: The correct term is Message Queue and not Messaging Queue.

File: Core/Utilities/RabbitMQConnector.py [hunk 4]
```diff
@@ -0,0 +1,154 @@
+"""
+Class for management of RabbitMQ connections
+"""
+
+__RCSID__ = "$Id$"
+
+import json
+import stomp
+import threading
+from MessagingQueueConnector import MessagingQueueConnector as MQConnector
```
Comment (atsareg) [inline Core/Utilities/RabbitMQConnector.py]: May be just make the base class MQConnector. MQ is now a well known acronym

**Comment by atsareg** [comment]
It's getting much better. I would like that we discuss the following organization of the messaging framework: Messaging Service, RabbitMQ in our case, is itself a third party service for DIRAC. As such it should be described as yet another Resource type in the CS and the Connectors are clients of those services pretty much like Catalog Clients or Storage Clients in DIRAC.Resources.

File: Core/Utilities/RabbitMQConnector.py [hunk 5]
```diff
@@ -0,0 +1,154 @@
+"""
+Class for management of RabbitMQ connections
+"""
+
+__RCSID__ = "$Id$"
+
+import json
+import stomp
+import threading
+from MQConnector import MQConnector
```
Comment (fstagni) [inline Core/Utilities/RabbitMQConnector.py]: from DIRAC.Core.Utilities.MQConnector import MQConnector

File: Core/Utilities/RabbitMQConnector.py [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+
+    return S_OK( msg )
+
+  def unsetupConnection( self ):
+    """
+    Disconnects from the RabbitMQ server
+    """
+    try:
+      self.connection.disconnect()
+    except Exception, e:
```
Comment (fstagni) [inline Core/Utilities/RabbitMQConnector.py]: except Exception as e:

File: Core/Utilities/RabbitMQConnector.py [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+
+    # Read parameters from CS
+    for parameter in [ 'Host', 'Port', 'User', 'VH', 'ExchangeName' ]:
+      result = self.__setQueueParameter( system, queueName, parameter )
+      if not result[ 'OK' ]:
+        return result
+
+    result = gConfig.getOption( '/LocalInstallation/MessageQueueing/Password' )
+    if not result[ 'OK' ]:
+      return S_ERROR( 'Failed to get the password for RabbitMQ: %s' % result[ 'Message' ] )
```
Comment (fstagni) [inline Core/Utilities/RabbitMQConnector.py]: Maybe add also a check 
```
if not result['Value']
```

**Comment by fstagni** [comment]
I would say that the __setQueueParameter function can be put in the general class. Also, reading of common CS parameters like Host, Port, User and Password can probably be added there. I would also add a CS parameter for each host called "type" (whose value could be, e.g. "RabbitMQ", or "ActiveMQ", ecc.)

File: Core/Utilities/MQConnector.py [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+    gLogger.error( message )
+
+  # Rest of functions
+
+  def setQueueParameters( self, system, queueName ):
+    """
+    Reads the MessageQueue parameters from the CS and sets the appropriate variables in the class with their values
+    system is the DIRAC system where the queue works
+    queueName is the name of the queue
+    """
```
Comment (fstagni) [inline Core/Utilities/MQConnector.py]: @wkrzemien @miloszz 
On a second thought, after consultation, we agreed that this function should be partially re-written: in fact, the only parameter that should be get from the CS is the password. All the others should be passed in, e.g. by  ConsumerModule, that should discover them.

Comment (fstagni) [inline Core/Utilities/MQConnector.py]: Reminder: this still needs to be updated.

**Comment by zmathe** [comment]
It is recreated #2974

**Pull request closed without merging**